Change isolation rules for multiple bridge networks

aboch opened the pull request:
- from subnet to interface

Related to https://github.com/docker/docker/issues/18297

Signed-off-by: Alessandro Boch <aboch@docker.com>

mavenugo commented on drivers/bridge/bridge.go:
```diff
@@ -300,7 +300,8 @@ func (n *bridgeNetwork) isolateNetwork(others []*bridgeNetwork, enable bool) err
 
-		if thisV6 != nil && otherV6 != nil && !types.CompareIPNet(thisV6, otherV6) {
+		// TODO: Enable when full ipv6 table infrastructure is available
+		if false && thisV6 != nil && otherV6 != nil && !types.CompareIPNet(thisV6, otherV6) {
```
As discussed, should we even completely avoid adding ipv4/v6 addresses to the iptables filter rules and instead use @mrjana 's idea of using device based filters ?

aboch replied:
Yes, that would be cleaner.
Although this PR's change is meant to unblock multiple network creation which is now impossible if  user assigns a V6 subnets.

Let me check if I can get the interface based isolation rule in place and working soon.

aboch commented:
@mavenugo 
Following your suggestion on @mrjana's method, I changed the inter-network isolation rules. Also fixed a bug in iptables.Exists() function (old one which causes duplicate rules) and adjust ICC rule to fit in the new logic.

Tested manually and passes docker integ-test-cli

PTAL

aboch commented:
@mavenugo @mrjana 
I also refactored the calls around `ipatbels.Raw()` to remove some repetitive code.

PTAL when you get a chance.
Thanks.

aboch commented:
@mrjana
As per your suggestion I created a new filter chain for the isolation rules.
PTAL when you get a chance. Thanks.

mrjana commented on drivers/bridge/setup_ip_tables.go:
```diff
@@ -11,20 +11,21 @@ import (
 const (
-	DockerChain = "DOCKER"
+	DockerChain    = "DOCKER"
+	IsolationChain = "ISOLATION"
```
Can we name it `DOCKER-ISOLATION`? Just to follow a convention, we are already calling overlay chain as `DOCKER-OVERLAY` so people looking at it will know that this is for docker

aboch replied:
Ok

mrjana commented on drivers/bridge/setup_ip_tables.go:
```diff
@@ -36,10 +37,26 @@ func setupIPChains(config *configuration) (*iptables.ChainInfo, *iptables.ChainI
+		}
+	}()
+
+	isolationChain, err := iptables.NewChain(IsolationChain, iptables.Filter, hairpinMode)
```
We don't need to pass `hairpinMode` to this chain as this is a filter only chain, it doesn't really matter if it is hairpin mode or not.

aboch replied:
I know, but the iptables package api is like that, and discards it anyway for filter table chains.
Passing a "true/false" would raise the same questions.

aboch replied:
I will pass a "false"

mrjana replied:
You can pass false. It is only applicable to `Nat` table so it is optional argument. The current way creates confusion for anybody reading the code who are going to think that hairpin mode means something to filter table as well. You can also remove it from the DOCKER filter table chain also if you would like consistency.

mrjana commented on drivers/bridge/setup_ip_tables.go:
```diff
@@ -36,10 +37,26 @@ func setupIPChains(config *configuration) (*iptables.ChainInfo, *iptables.ChainI
 	filterChain, err := iptables.NewChain(DockerChain, iptables.Filter, hairpinMode)
 	if err != nil {
-		return nil, nil, fmt.Errorf("Failed to create FILTER chain: %s", err.Error())
+		return nil, nil, nil, fmt.Errorf("Failed to create FILTER chain: %v", err)
```
error string to start in lower case

aboch replied:
Will take care of it

mrjana commented on drivers/bridge/setup_ip_tables.go:
```diff
@@ -36,10 +37,26 @@ func setupIPChains(config *configuration) (*iptables.ChainInfo, *iptables.ChainI
+
+	isolationChain, err := iptables.NewChain(IsolationChain, iptables.Filter, hairpinMode)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("Failed to create FILTER isolation chain: %v", err)
```
error string to start in lower case

mrjana commented:
LGTM

mavenugo commented:
@aboch changes LGTM.
Given that the changes impact bridge networking, I would like to do some additional testing especially the compatibility with previous versions.

mavenugo commented:
related to https://github.com/docker/docker/issues/18354

mavenugo commented:
LGTM
